gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -599,6 +734,689 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
+        return CE_None;
+    }
+
+    poDS->bNoDataSet = true;
+    poDS->dfNoDataValue = dfNoData;
+
+    poDS->bNoDataChanged = true;
+
+    bNoDataSet = true;
+    dfNoDataValue = dfNoData;
+    return CE_None;
+}
+
+/************************************************************************/
+/*                        DeleteNoDataValue()                           */
+/************************************************************************/
+
+CPLErr FITSRasterBand::DeleteNoDataValue()
+
+{
+    poDS->LoadGeoreferencingAndPamIfNeeded();
+
+    if( !poDS->bNoDataSet )
+        return CE_None;
+
+    poDS->bNoDataSet = false;
+    poDS->dfNoDataValue = -9999.0;
+
+    poDS->bNoDataChanged = true;
+
+    bNoDataSet = false;
+    dfNoDataValue = -9999.0;
+    return CE_None;
+}
+
+/************************************************************************/
+/*                     LoadGeoreferencingAndPamIfNeeded()               */
+/************************************************************************/
+
+void FITSDataset::LoadGeoreferencingAndPamIfNeeded()
+
+{
+    int status = 0;
+    int bitpix;
+    double dfScale, dfOffset;
+    double crpix1, crpix2, crval1, crval2, cdelt1, cdelt2, pc[4], cd[4];
+    double aRadius, cRadius, invFlattening = 0.0;
+    char target[16], ctype[9];
+    std::string pszGeogName, pszDatumName, projName;
+
+    const double PI = std::atan(1.0)*4;
+    const double DEG2RAD = PI / 180.;
+
+    if( !m_bReadGeoTransform && !m_bLoadPam )
+        return;
+
+/* -------------------------------------------------------------------- */
+/*      Get the transform from the FITS file.                           */
+/* -------------------------------------------------------------------- */
+    if( m_bReadGeoTransform )
+    {
+        m_bReadGeoTransform = false;
+
+        fits_read_key(hFITS, TSTRING, "OBJECT", target, nullptr, &status);
+        if( status )
+        {
+            strncpy(target, "Undefined", 10);
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "OBJECT keyword is missing");
+            status = 0;
+        }
+
+        pszGeogName.assign("GCS_");
+        pszGeogName.append(target);
+        pszDatumName.assign("D_");
+        pszDatumName.append(target);
+
+        fits_read_key(hFITS, TDOUBLE, "A_RADIUS", &aRadius, nullptr, &status);
+        if( status )
+        {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "No Radii keyword available, metadata will not contain DATUM information.");
+            status = 0;
+        } else {
+            fits_read_key(hFITS, TDOUBLE, "C_RADIUS", &cRadius, nullptr, &status);
+            if( aRadius != cRadius )
+            {
+                invFlattening = aRadius / ( aRadius - cRadius );
+            }
+        }
+
+        /* Waiting for linear keywords standardization only deg ctype are used */
+        /* Check if WCS are there */
+        fits_read_key(hFITS, TSTRING, "CTYPE1", ctype, nullptr, &status);
```

Comment:
how can be sure that ctype[9] won't be overflown in case of hostile/corrupted images ? Does fits_read_key() guarantee that the maximum number of bytes written will be 8 + a zero terminator ?

Reply:
what if CTYPE1 doesn't exist as a key ? Will ctype[0] be set to 0 ?

Reply:
CFITSIO error codes cover a large number of errors (https://heasarc.nasa.gov/docs/software/fitsio/quick/node26.html). CTYPE is the mandatory keyword to have a coordinate system defined: so if `status != 0` I just skip the definition of georeferentiation and projection. Yet, I have added one more check of `status` on CRPIX/CRVAL keywords: again, if they are not there or the `read` function throws an error the georef is skipped.

Reply:
That doesn't completely reassure me there is no risk of overflow of ctype[9] in case a hostile file would have been created on purpose with a CTYPE1 value with 80 characters. We try in GDAL to defend against that.

Reply:
Will a `ctype[80]` reassure you?

Reply:
> Will a `ctype[80]` reassure you?

yes. Wondering if it shouldn't be [81] for the null terminator actually ? (same for target[] then)

Reply:
ok, let me check about the null terminator and I will commit the changes.

Reply:
Fixed in https://github.com/OSGeo/gdal/pull/1298/commits/47acb1256b6adf26157106b31e2626dd3fc210b2